### ISSUE TITLE
Forward reference in generic type argument causes missing members

### ISSUE THREAD

**Issue by Pyprohly** [microsoft/pyright#2237]
#### Describe the bug

I have a forward reference in a generic type argument referring to a class whose base class is the class using the forward reference. This situation causes Pyright to not inherit members when subclassing the class using the forward reference.

#### To Reproduce

```python
from __future__ import annotations
from typing import TypeVar, Generic, Sequence

class Comment: ...
class Submission: ...

T = TypeVar('T')
TChild = TypeVar('TChild')

class GeneralTreeNode(Generic[T, TChild]):
    def __init__(self,
        value: T,
        children: Sequence[TChild],
):
        self.value = value
        self.children = children

class CommentSubtreeTreeNode(GeneralTreeNode[T, 'CommentTreeNode']):
    pass

class CommentTreeNode(CommentSubtreeTreeNode[Comment]):
    pass

class SubmissionTreeNode(CommentSubtreeTreeNode[Submission]):
    pass

c = CommentTreeNode(Comment(), [])
c.value
c.children
```

Error preview:

```text
error: Expected no arguments to "CommentTreeNode" constructor (reportGeneralTypeIssues)
error: Cannot access member "value" for type "CommentTreeNode"
    Member "value" is unknown (reportGeneralTypeIssues)
error: Cannot access member "children" for type "CommentTreeNode"
    Member "children" is unknown (reportGeneralTypeIssues)
```

The code successfully type checks in Mypy.

#### My Environment

- Python: 3.9.6
- Pyright: 1.1.163
- Mypy: 0.910

**Comment by erictraut** [comment]
Forward references work fine, but you've created a circular reference here that pyright cannot handle. `CommentSubtreeTreeNode` derives from `GeneralTreeNode[T, CommentTreeNode]`, but `CommentTreeNode` derives from `CommentSubTreeNode`. 

Mypy is able to deal with a few cases of circular references that pyright cannot. This is one such case. This is due to the underlying architecture of the two type checkers. Mypy is a multi-pass analyzer, and it analyzes the entire document multiple times until all types converge. Pyright is a just-in-time (aka "lazy") analyzer that visits all nodes just once but in any order necessary to resolve symbol types. This is what makes pyright so much faster.

You'll need to eliminate this circular dependency if you want your code to work with pyright.

**Comment by Pyprohly** [comment]
That’s unfortunate. The circular dependency seems necessary to what I’m trying to model though.

Before using these type checkers I would rewrite complicated typing situations like this in C# as a benchmark to double check that the code I wrote made sense, and translating the above code to C# does produces valid code, FWIW.

<details>
    <summary>(C# code)</summary>

```csharp
using System;
using System.Collections.Generic;

namespace console
{
    public class Comment { }
    public class Submission { }

    public class GeneralTreeNode<T, TChild>
    {
        public T value;
        public IList<TChild> children;

        public GeneralTreeNode(T value, IList<TChild> children)
        {
            this.value = value;
            this.children = children;
        }
    }

    public class CommentSubtreeTreeNode<T> : GeneralTreeNode<T, CommentTreeNode>
    {
        public CommentSubtreeTreeNode(T value, IList<CommentTreeNode> children) :
            base(value, children)
        { }
    }
    public class CommentTreeNode : CommentSubtreeTreeNode<Comment>
    {
        public CommentTreeNode(Comment value, IList<CommentTreeNode> children) :
            base(value, children)
        { }
    }
    public class SubmissionTreeNode : CommentSubtreeTreeNode<Submission>
    {
        public SubmissionTreeNode(Submission value, IList<CommentTreeNode> children) :
            base(value, children)
        { }
    }

    public class MainClass
    {
        public static void Main(string[] args)
        {
            IList<CommentTreeNode> children = new List<CommentTreeNode>();
            var c = new CommentTreeNode(new Comment(), children);
            _ = c.value;
            _ = c.children;
        }
    }
}
```
</details>

Will we ever get multi-pass in Pyright? Maybe it could be enabled though a switch down the road if the slowdown is significant.

**Comment by erictraut** [comment]
Multi-pass analysis would require an entirely different architecture with no commonality to the current pyright. Very early versions of pyright used a multi-pass approach, but I did a complete rewrite about two years ago to improve speed. Performance and lazy evaluation are especially important for a type analyzer that forms the basis for a language server because user expect quick responses when interactively editing (e.g. completion suggestions, etc.).
